compiler/rustc_typeck/src/check/cast.rs

```diff
@@ -538,6 +540,17 @@ impl<'a, 'tcx> CastCheck<'tcx> {
                 }
                 err.emit();
             }
+            CastError::ForeignNonExhaustiveAdt => {
+                make_invalid_casting_error(
+                    fcx.tcx.sess,
+                    self.span,
+                    self.expr_ty,
+                    self.cast_ty,
+                    fcx,
+                )
+                .note("cannot cast a non-exhaustive enum defined in another crate")
```

Comment:
nitpick: I think this more "an enum with a non-exhaustive variant" than "a non-exhaustive enum", right?

(The latter would be `#[non_exhaustive] pub enum Foo { ... }`.)

Reply:
(aside: It would be good to know whether the `#[non_exhaustive] pub enum Foo { ... }` case is already rejected. If it isn't, it should be.)

Reply:
Isn't that case one of the main uses of `#[non_exhaustive]`? (Unless I'm misunderstanding what you mean by "rejected".)

Reply:
@camelid "Rejected" as in "`as` cross-crate should also be rejected when the *enum* itself is non-exhaustive", like this PR is making be the behaviour be for when any *variant* is non-exhaustive.

I don't see any obvious test for it in https://github.com/rust-lang/rust/tree/master/src/test/ui/rfc-2008-non-exhaustive, at least.  Would be a good thing to try after this PR goes in.